soroban-env-host/src/native_contract/token/contract.rs

```diff
@@ -93,12 +112,92 @@ pub struct Token;
 
 #[contractimpl]
 impl TokenTrait for Token {
-    fn initialize(e: &Host, admin: Identifier, metadata: Metadata) -> Result<(), Error> {
-        if has_administrator(&e)? {
+    fn init_wrap(e: &Host, metadata: ClassicMetadata) -> Result<(), Error> {
+        if has_metadata(&e)? {
             return Err(Error::ContractError);
         }
+
+        match metadata.clone() {
+            ClassicMetadata::Native => {
+                let contract_id = BytesN::<32>::try_from_val(e, e.get_current_contract()?)?;
+                let xlm_contract_id =
+                    BytesN::<32>::try_from_val(e, e.get_contract_id_from_asset(Asset::Native)?)?;
+                if contract_id != xlm_contract_id {
+                    return Err(Error::ContractError);
+                }
+
+                write_metadata(&e, Metadata::Native)?;
+                //No admin for the Native token
+            }
+            ClassicMetadata::AlphaNum4(asset) => {
+                let contract_id = BytesN::<32>::try_from_val(e, e.get_current_contract()?)?;
+
+                //TODO: Better way to do this?
+                let mut code4 = [0u8; 4];
```

Comment:
Can we factor this (i.e. code/issuer extraction) out into a helper function and reuse it below?